src/main/java/com/rultor/agents/docker/StartsDockerDaemon.java

```diff
@@ -0,0 +1,148 @@
+ * are met: 1) Redistributions of source code must retain the above
+ * copyright notice, this list of conditions and the following
+ * disclaimer. 2) Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following
+ * disclaimer in the documentation and/or other materials provided
+ * with the distribution. 3) Neither the name of the rultor.com nor
+ * the names of its contributors may be used to endorse or promote
+ * products derived from this software without specific prior written
+ * permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT
+ * NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+ * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.rultor.agents.docker;
+
+import com.github.dockerjava.api.DockerClient;
+import com.github.dockerjava.api.command.CreateContainerResponse;
+import com.github.dockerjava.api.model.ExposedPort;
+import com.github.dockerjava.api.model.Ports;
+import com.github.dockerjava.core.DockerClientBuilder;
+import com.github.dockerjava.core.command.BuildImageResultCallback;
+import com.rultor.agents.shells.PfShell;
+import com.rultor.spi.Profile;
+import java.io.File;
+import java.io.IOException;
+import java.io.StringWriter;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.concurrent.ConcurrentHashMap;
+import org.codehaus.plexus.util.IOUtil;
+
+/**
+ * Starts a Docker Container containing a Docker daemon and SSHD.
+ * @author Armin Braun (me@obrown.io)
+ * @version $Id$
+ * @since 1.63
+ */
+public final class StartsDockerDaemon implements AutoCloseable {
+
+    /**
+     * Profile used.
+     */
+    private final transient Profile profile;
+
+    /**
+     * Docker client.
+     */
+    private final transient DockerClient client;
+
+    /**
+     * Docker Containers created.
+     */
+    private final transient Collection<CreateContainerResponse> containers;
+
+    /**
+     * Ctor.
+     * @param prof Current Profile
+     */
+    public StartsDockerDaemon(final Profile prof) {
+        this.profile = prof;
+        this.client = DockerClientBuilder.getInstance().build();
+        this.containers = Collections.newSetFromMap(
+            // @checkstyle MagicNumber (1 line)
+            new ConcurrentHashMap<CreateContainerResponse, Boolean>(1, 0.9f, 1)
```

Comment:
@mkordas yea exactly, my reasoning simply was that  this way we're safer from interruptions in the `close`. Admittedly it's a prime case of premature generalization at this point though ...,
But since it's also somewhat more concise than a synchronized set + synchronizing on iteration, I'd go with it.